apps/studio/src/lib/db/clients/postgresql.ts

```diff
@@ -1358,10 +1359,46 @@ function configDatabase(server: { sshTunnel: boolean, config: IDbConnectionServe
     }
   }
 
+  // For Redshift Only -- IAM authentication and credential exchange
```

Comment:
Could we move this logic into a file like `lib/db/authentication/amazon.ts`? That way we can re-use it for other Amazon IAM-authenticated databases.

Reply:
Then we can do something like:

```
const { tempUser, tempPassword } = iamAuthenticate(server.config.options)
```

Reply:
Yeah that makes sense. One caveat to using `amazon.ts` is that this particular credential exchange is specific to Redshift. RDS has a slightly different approach from what I recall. I'll move this out into its own file and make sure it's obvious this is a Redshift-specific operation.

Reply:
After digging into this a little more, the credential processes are slightly different, though all could conform to a single interface. I'm thinking that's a future enhancement since RDS isn't a formal database option right now.

I've created a `RedshiftCredentialResolver` which will handle exchanging the credentials. While implementing the password callback for `pg` I noticed it was being invoked far more frequently than needed, which could result in throttling from AWS.

At the risk of over-blowing it a little, I created the resolver using a singleton pattern to prevent accidentally having two of these in runtime retrieving independent credentials, and a light-weight caching mechanism that will return good credentials if invoked frequently.